### PR TITLE
Update `upload-artifact` and `download-artifact` actions

### PR DESCRIPTION
**Pull request by lunaris**
v2 of `actions/upload-artifact` and `actions/download-artifact` will be deprecated on June 30, 2024. This commit bumps the versions of these actions to v4. The main breaking change between v2/v3 and v4 is that artifacts are now immutable, so multiple uploads to artifacts with the same name will fail by default. In this case there are two options -- use different names, and recollect the artifacts later with a modified `download-artifact` instantiation, or pass `overwrite: true` to calls which will happen later to first delete the artifact and create a new one in its place. It _appears_ that we only fall foul of repeated uploads in one place, and that for this case we can use `overwrite`, so that is the approach this commit takes.

### CODE REVIEW

File: .github/workflows/ci-run-test.yml [hunk 1]
```diff
@@ -373,9 +373,10 @@ jobs:
         shell: bash
         run: mkdir status; echo "OK" > "${{ env.TEST_STATUS_FILE }}"
       - name: Register test successful.
-        uses: actions/upload-artifact@v3
+        uses: actions/upload-artifact@v4
         with:
           path: ${{ env.TEST_STATUS_FILE }}
+          overwrite: true
```
Comment (lunaris) [inline .github/workflows/ci-run-test.yml:379]: I think here we are making a later write to the test status file to indicate success, which means it's OK to overwrite the previous instance?

**Review by julienp — APPROVED** [review]

File: .github/workflows/ci-run-test.yml (hunk 1 above)
Comment (julienp) [reply .github/workflows/ci-run-test.yml:379]: Compute says no

😭

**Comment by justinvp** [comment]
All the tests passed in the most recent merge attempt, but the `test-collect-coverage` job failed. I haven't dug into it, but I am wondering if it's a legit issue related to artifact downloading.

https://github.com/pulumi/pulumi/actions/runs/9704560770/job/26785864891#step:2:126

**Comment by julienp** [comment]
```
##[debug]if test -z "$(echo {
##[debug]  "unit-test": {
##[debug]    "result": "success",
##[debug]    "outputs": {}
##[debug]  },
##[debug]  "integration-test": {
##[debug]    "result": "success",
##[debug]    "outputs": {}
##[debug]  },
##[debug]  "acceptance-test-win": {
##[debug]    "result": "success",
##[debug]    "outputs": {}
##[debug]  },
##[debug]  "acceptance-test-macos": {
##[debug]    "result": "success",
##[debug]    "outputs": {}
##[debug]  }
##[debug]} | jq -r ''.[] | .outputs | .result'' | grep -v ''success'')"; then
##[debug]    echo "Tests OK!"
##[debug]else
##[debug]    echo "Test Failure.. skipping CodeCov upload."
##[debug]    exit 1
##[debug]fi
```

I think we can't have the multiline in the `if` here,

We could instead split it into a variable and try to get to something like

```bash
JSON = '{
  "unit-test": {
  ...
'

if test -z "$echo $JSON | jq -r ''.[] | .outputs | .result'' | grep -v ''success'')"; then
    echo "Tests OK!"
else
    echo "Test Failure.. skipping CodeCov upload."
    exit 1
fi
```

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -444,14 +445,11 @@ jobs:
...
-        uses: actions/download-artifact@v3
-        with:
-          path: status
-      # Check that there are no incomplete tests.
+      # Check that there are no failed tests.
       - name: Check tests completed successfully.
         run: |
-          if test -z "$(grep -rl 'incomplete' ./status)"; then
+          # turn this into an if
+          if test -z "$(echo ${{ toJson(needs) }} | jq -r '.[] | .outputs | .result' | grep -v 'success')"; then
```
Comment (julienp) [inline .github/workflows/ci.yml:452]:
```suggestion
          JSON='${{ toJson(needs) }}'
          if test -z "$(echo $JSON | jq -r '.[] | .outputs | .result' | grep -v 'success')"; then
```

File: .github/workflows/ci-build-binaries.yml [hunk 3]
```diff
@@ -130,7 +130,7 @@ jobs:
           cat .goreleaser.yml \
             | go run github.com/t0yv0/goreleaser-filter@v0.3.0 -goos ${{ inputs.os }} -goarch ${{ inputs.arch }} \
             | goreleaser release -f - -p 5 --skip-validate --clean --snapshot
-      - uses: actions/upload-artifact@v2
+      - uses: actions/upload-artifact@v4
```
Comment (julienp) [inline .github/workflows/ci-build-binaries.yml:133]: I think this needs `overwrite: true`

File: .github/workflows/ci.yml (hunk 2 above)
Comment (julienp) [reply .github/workflows/ci.yml:452]: In the debug output in https://github.com/pulumi/pulumi/actions/runs/9704560770/job/26785864891#step:2:126 we have two single quotes around the strings in the test: `jq -r ''.[] | .outputs | .result''`, I am not sure what's up with that.

File: .github/workflows/ci-build-binaries.yml (hunk 3 above)
Comment (julienp) [reply .github/workflows/ci-build-binaries.yml:133]: Pushed here https://github.com/pulumi/pulumi/pull/16485/commits/ab9be11e8fb9abb57d2ccf686129d61012a7951c

**Review by tgummerer — APPROVED** [review]

**Pull request merged**